File: src/main/java/core/reader/FileReader.java
package core.reader;

import core.exceptions.IllegalSyntaxException;
import core.exceptions.SCMFileNotFoundException;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PushbackReader;
import java.util.ArrayList;
import java.util.List;

public class FileReader extends Reader {

  public List<Object> read(File file) {
    try {
      reader = new PushbackReader(new BufferedReader(new java.io.FileReader(file)), 2);
    } catch (FileNotFoundException e) {
      throw new SCMFileNotFoundException(file.getPath());
    }
    List<Object> tokens = new ArrayList<>();
    try {
      int read;
      while ((read = reader.read()) != -1) {
        reader.unread(read);
        Object token = nextToken();
        if (DOT.equals(token)) {
          throw new IllegalSyntaxException("read: illegal use of '.'");
        }
        if (token != null) {
          tokens.add(token);
        }
      }
    } catch (IOException e) {
      e.printStackTrace();
    } finally {
      try {
        reader.close();
      } catch (IOException ignore) {
      }
    }
    return tokens;
  }
}

File: src/main/java/core/reader/FileReader.kt
package core.reader

import java.io.*

class FileReader : Reader() {

    fun read(file: File): List<Any> = run {
        reader = PushbackReader(BufferedReader(java.io.FileReader(file)), 1)
        reader.use {
            ArrayList<Any>().apply {
                var i = reader.read()
                while (i != -1) {
                    reader.unread(i)
                    nextToken()?.let { add(it) }
                    i = reader.read()
                }
            }
        }
    }
}
